financial-templates-lib/price-feed/utils.js

```diff
@@ -0,0 +1,130 @@
+const lodash = require("lodash");
+const assert = require("assert");
+
+// Downloads blocks and caches them for certain time into the past.
+// Allows some in memory searches to go from timestamp to block number.
+exports.BlockHistory = (web3, blocks = []) => {
+  assert(web3, "requires web3");
+
+  // Check if we have downloaded a block by number
+  function has(number) {
+    return blocks.find(block => block.number === number);
+  }
+
+  // Used internally, but will insert a block into cache sorted by timestamp
+  function insert(block) {
+    const index = lodash.sortedIndexBy(blocks, block, "timestamp");
+    blocks.splice(index, 0, block);
+    return blocks;
+  }
+
+  // Get the block with the closest match to timestamp
+  function getClosestTime(timestamp) {
+    // const index = lodash.sortedIndexBy(blocks,{timestamp},'timestamp')
+    const index = lodash.sortedIndexBy(blocks, { timestamp }, x => {
+      return x.timestamp;
+    });
+    return blocks[index];
+  }
+
+  // Call to update cache with blocks between sometime in the past from a block number
+  // Block number is optional, if not specified will default to latest block number.
+  async function fetchBetween(endTime, blockNumber) {
```

Comment:
hmm yea, like in the comment above, this is really meant to be private, but is exposed for the option of testing it independently ( though there arent tests currently for it).  I could just internalize these functions since im not testing them and just call it a day.